<a href="https://colab.research.google.com/github/susvej/bg_ontology/blob/main/Practice_SPARQL_with_the_Boardgame_Knowledge_Graph.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practice SPARQL
You already know the basics of Sparql from various online tutorials. Now you need to practice, practice, practice.

Make a copy of this notebook and start practicing. Feel free to email additional practice problems (and your suggested solutions) to susanne+kgpractice@vejdemo.se.

#The board game knowledge graph
Read about the board game KG [here](https://github.com/susvej/bg_ontology/blob/main/README.md). You can explore it will sparql queries below.

#Structure of this Colab
We will be using a python library called rdflib. If you are not interested in python, you can ignore that part, and only focus on the sparql queries.
All python housekeeping necessary to make the queries work will be in cells marked "Python"

In [ ]:
# Python housekeeping
# Install the rdflib library

!pip install rdflib


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 565.1/565.1 kB 25.7 MB/s eta 0:00:00


In [ ]:
# Python housekeeping
# Import Graph
from rdflib import Graph, Namespace, URIRef, RDF, RDFS, OWL, Literal
from rdflib.plugins.sparql import prepareQuery
from rdflib import URIRef


In [ ]:

# Initialize the graph
g = Graph()

# Parse the TTL files
g.parse("https://raw.githubusercontent.com/susvej/bg_ontology/refs/heads/main/bgg_main.ttl", format="ttl")

bgg = Namespace("https://raw.githubusercontent.com/susvej/bg_ontology/")
g.namespace_manager.bind("bgg", bgg, override=True)

# Print number of triples
print(f"Graph has {len(g)} triples.")

# Print first 20 triples using qname for URIs, quoted strings for literals
# This is just some random data.
for s, p, o in list(g)[:20]:
    s_str = g.qname(s) if isinstance(s, URIRef) else str(s)
    p_str = g.qname(p) if isinstance(p, URIRef) else str(p)
    if isinstance(o, URIRef):
        o_str = g.qname(o)
    elif isinstance(o, Literal):
        o_str = f'"{o}"'
    else:
        o_str = str(o)
    print(s_str, p_str, o_str)

Graph has 83348 triples.
bgg:5233 bgg:hasMinRecAge "12"
bgg:156138 bgg:hasMaxPlayers "6"
bgg:223855 bgg:hasMechanic bgg:Card_Drafting
bgg:28185 bgg:hasCategory bgg:World_War_I
bgg:4741 bgg:hasCreator "Mike McVey, Brian Snoddy, Matt Wilson"
bgg:124380 bgg:hasMechanic bgg:Partnerships
bgg:141090 bgg:hasURL "https://boardgamegeek.com/boardgame/141090/kohle-kolonie"
bgg:124356 bgg:hasMaxPlayers "4"
bgg:3570 bgg:hasName "Chicken Cha Cha Cha"
bgg:1552 bgg:hasMechanic bgg:Hand_Management
bgg:14683 bgg:hasCategory bgg:Medieval
bgg:178896 bgg:hasMechanic bgg:Hex-and-Counter
n258485e9b0c849368c0bc408002e611fb5 rdf:first bgg:easy
bgg:148759 bgg:hasMinPlayers "2"
bgg:7086 bgg:hasMaxPlayers "2"
bgg:149869 bgg:hasCategory bgg:Deduction
bgg:134157 bgg:hasMechanic bgg:Hand_Management
bgg:3893 bgg:hasCreator "Jon M. Tuffley"
bgg:56320 bgg:hasMechanic bgg:Card_Drafting
bgg:72991 bgg:hasMechanic bgg:Pattern_Building


In [ ]:
# Python housekeeping, but run this cell even if you have no
# interest in Python, and look at the results to explore the graph

# g.qname gives prefix:localpart rather than whole IRI
qname = g.qname

print("CLASSES:")
for s in g.subjects(RDF.type, RDFS.Class):
    print(" -", qname(s))

print("\nPROPERTIES:")
for s in g.subjects(RDF.type, RDF.Property):
    print(" -", qname(s))
for s in g.subjects(RDF.type, OWL.ObjectProperty):
    print(" -", qname(s))
for s in g.subjects(RDF.type, OWL.DatatypeProperty):
    print(" -", qname(s))

print("\nOBJECT PROPERTY RELATIONSHIPS:")
for prop in set(g.subjects(RDF.type, OWL.ObjectProperty)):
    dom = list(g.objects(prop, RDFS.domain))
    ran = list(g.objects(prop, RDFS.range))
    print(f" - {qname(prop)}")
    if dom:
        print(f"   ↳ Domain: {', '.join(qname(d) for d in dom)}")
    if ran:
        print(f"   ↳ Range:  {', '.join(qname(r) for r in ran)}")

print("\nDATATYPE PROPERTY RELATIONSHIPS:")
for prop in set(g.subjects(RDF.type, OWL.DatatypeProperty)):
    dom = list(g.objects(prop, RDFS.domain))
    ran = list(g.objects(prop, RDFS.range))
    print(f" - {qname(prop)}")
    if dom:
        print(f"   ↳ Domain: {', '.join(qname(d) for d in dom)}")
    if ran:
        print(f"   ↳ Range:  {', '.join(qname(r) for r in ran)}")

#Todo(Sus, April 5, 2025, Note that there is an error where datatype property hasMentalLoad takes a class range.
#Non-urgent fix, since there is no data of that type yet

CLASSES:
 - bgg:Category
 - bgg:Game
 - bgg:Mechanic
 - bgg:MentalLoad
 - bgg:Player
 - bgg:Size

PROPERTIES:
 - bgg:hasCategory
 - bgg:hasID
 - bgg:hasMechanic
 - bgg:hasSize
 - bgg:hasURL
 - bgg:hasOwnershipOf
 - bgg:isCategoryOf
 - bgg:isMechanicOf
 - bgg:isOwnedBy
 - bgg:likesCategory
 - bgg:likesMechanic
 - bgg:hasBestNumPlayers
 - bgg:hasComment
 - bgg:hasCreator
 - bgg:hasGeekRating
 - bgg:hasMaxGameTime
 - bgg:hasMaxPlayers
 - bgg:hasMaxRecAge
 - bgg:hasMentalLoad
 - bgg:hasMinGameTime
 - bgg:hasMinPlayers
 - bgg:hasMinRecAge
 - bgg:hasName
 - bgg:hasPublisher
 - bgg:hasRating

OBJECT PROPERTY RELATIONSHIPS:
 - bgg:isCategoryOf
   ↳ Domain: bgg:Category
   ↳ Range:  bgg:Game
 - bgg:likesMechanic
   ↳ Domain: bgg:Player
   ↳ Range:  bgg:Mechanic
 - bgg:isOwnedBy
   ↳ Domain: bgg:Game
   ↳ Range:  bgg:Player
 - bgg:hasOwnershipOf
   ↳ Domain: bgg:Player
   ↳ Range:  bgg:Game
 - bgg:isMechanicOf
   ↳ Domain: bgg:Mechanic
   ↳ Range:  bgg:Game
 - bgg:likesCategory
   ↳ Domain: bgg:

In [ ]:
# Explore the list of Categories and Mechanics

print("MECHANICS:")
for s in g.subjects(RDF.type, bgg.Mechanic):
    print(" -", qname(s))

print("\n\nCATEGORIES:")
for s in g.subjects(RDF.type, bgg.Category):
    print(" -", qname(s))


#Todo(Sus, May 15, 2025, note error of ns1:_Play - check out
#Non-urgent fix

MECHANICS:
 - bgg:Acting
 - bgg:Action_Point_Allowance_System
 - bgg:Action_or_Movement_Programming
 - bgg:Area-Impulse
 - bgg:Area_Control_or_Area_Influence
 - bgg:Area_Enclosure
 - bgg:Area_Movement
 - bgg:AuctionorBidding
 - bgg:BettingorWagering
 - bgg:Campaign_or_Battle_Card_Driven
 - bgg:Card_Drafting
 - bgg:Chit-Pull_System
 - bgg:Co-operative_Play
 - bgg:Commodity_Speculation
 - bgg:Crayon_Rail_System
 - bgg:Deck_or_Pool_Building
 - bgg:Dice_Rolling
 - bgg:Grid_Movement
 - bgg:Hand_Management
 - bgg:Hex-and-Counter
 - bgg:Line_Drawing
 - bgg:Memory
 - bgg:Modular_Board
 - bgg:Paper-and-Pencil
 - bgg:Partnerships
 - bgg:Pattern_Building
 - bgg:Pattern_Recognition
 - bgg:Pick-up_and_Deliver
 - bgg:Player_Elimination
 - bgg:Point_to_Point_Movement
 - bgg:Press_Your_Luck
 - bgg:Rock-Paper-Scissors
 - bgg:Role_Playing
 - bgg:Roll_or_Spin_and_Move
 - bgg:RouteorNetwork_Building
 - bgg:Secret_Unit_Deployment
 - bgg:Set_Collection
 - bgg:Simulation
 - bgg:Simultaneous_Action_Selection


In [ ]:
# Python housekeeping
# Print the results prettier.
import pandas as pd
def pretty(results):
    """
    Converts RDFLib SPARQL SELECT query results into a pandas DataFrame.

    Parameters:
        results (rdflib.query.Result): The result of a SELECT query on an RDFLib Graph.

    Returns:
        pd.DataFrame: A DataFrame with variable names as columns and query results as rows.
    """
    data = [row.asdict() for row in results]
    return pd.DataFrame(data)

# Practice time!

In [ ]:
# Problem 1: Show me 10 party games.
# Python explanation:
# This is an example query - note that first we create a variable "myquery" that holds the output
# of a  prepareQuery() function. The prepareQuery was imported above: "from rdflib.plugins.sparql import prepareQuery"
# After that we give myquery as input to the function g.query(), and then give the output of that as input to the "pretty" function that prints it prettier.
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

    SELECT ?name WHERE {
        ?game bgg:hasCategory bgg:Party_Game .
        ?game bgg:hasName ?name .
    }
    ORDER BY ?name
    LIMIT 10
""")

# Run the query and print results
pretty(g.query(myquery))


,name
0,1000 Blank White Cards
1,25 Words or Less
2,27th Passenger: A Hunt On Rails
3,30 Seconds
4,A Fake Artist Goes to New York
5,A la carte
6,Activity
7,Activity 2
8,America
9,Anno Domini: Deutschland


In [ ]:
# @title Problem 2: Show names of 10 Games with a Geek Rating Above 7.0

# Prepare the SPARQL query
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

    SELECT ?name ?rating WHERE {
        ?game bgg:hasGeekRating ?rating .
        ?game bgg:hasName ?name .
        FILTER(?rating > 7.0)
    }
    ORDER BY DESC(?rating)
    LIMIT 10
""")
# (ROUND(?rating) as ?roundrating)
pretty(g.query(myquery))


In [ ]:
# Your Solution
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

    SELECT * WHERE
    {?s ?p ?o}
    LIMIT 3
""")

pretty(g.query(myquery))

,p,s,o
0,http://www.w3.org/1999/02/22-rdf-syntax-ns#type,https://raw.githubusercontent.com/susvej/bg_on...,https://raw.githubusercontent.com/susvej/bg_on...
1,http://www.w3.org/1999/02/22-rdf-syntax-ns#type,https://raw.githubusercontent.com/susvej/bg_on...,https://raw.githubusercontent.com/susvej/bg_on...
2,https://raw.githubusercontent.com/susvej/bg_on...,https://raw.githubusercontent.com/susvej/bg_on...,90


In [ ]:
# @title Default title text
#Problem 3: Count party Games with a Geek Rating Above 7.0

# Prepare the SPARQL query
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

    SELECT (COUNT(?game) as ?populargames) WHERE {
      ?game rdf:type bgg:Game;
      bgg:hasGeekRating ?rating;
      bgg:hasCategory bgg:Party_Game
      FILTER (?rating > 7)
    }
""")

pretty(g.query(myquery))


,populargames
0,15


In [ ]:
#Problem 4: Games Suitable for 3 to 8 Players
# Prepare the SPARQL query
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name ?minplay ?maxplay WHERE {
?game bgg:hasName ?name ;
bgg:hasMinPlayers ?minplay ;
bgg:hasMaxPlayers ?maxplay .
FILTER (?minplay >2 && ?maxplay <8 && ?maxplay>2)
    }
""")
#
pretty(g.query(myquery))

,name,minplay,maxplay
0,Die Macher,3,5
1,Klunker,3,5
2,Risk: Metal Gear Solid,3,5
3,Pirates of Nassau,3,5
4,Robin Hood,3,5
...,...,...,...
581,Sleuth,3,7
582,1817,3,7
583,Hearts,3,7
584,1880: China,3,7


In [ ]:
#Problem 5: Count games for more than 2 players, but does not have Pirate or pirates or pirating in the title
# Prepare the SPARQL query
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (COUNT(?name) as ?nonpirategames) {
  ?game rdf:type bgg:Game;
  bgg:hasName ?name;
  bgg:hasMinPlayers ?minplayers
  FILTER(?minplayers > 2 && !regex(?name, "pirate", "i"))
}
""")
#
pretty(g.query(myquery))

,nonpirategames
0,774


In [ ]:
#Problem 5: 5 Games with 'Card Drafting' Mechanic in the Fantasy Category, ordered Z to A
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name WHERE {
  ?game bgg:hasName ?name;
  bgg:hasMechanic bgg:Card_Drafting;
  bgg:hasCategory bgg:Fantasy

}
ORDER BY DESC(?name)
LIMIT 5
""")
#
pretty(g.query(myquery))

,name
0,Wizards of the Wild: Deluxe Edition
1,Wizards of the Wild
2,Wicked Witches Way
3,Villagers & Villains
4,Viceroy


In [ ]:
#Problem 6: Card games that 2 people can play under 30 minutes. Name the resulting columnd "GameName" and "GameTime"

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (?name as ?GameName) (?maxtime as ?GameTime) WHERE {
  ?game bgg:hasName ?name;
  bgg:hasMinPlayers 2;
  bgg:hasMaxGameTime ?maxtime
  FILTER(?maxtime <= 30)

}
ORDER BY DESC(?name)
LIMIT 5
""")
#
pretty(g.query(myquery))


,GameName,GameTime
0,oddball Äeronauts,20
1,duck! duck! Go!,20
2,Zoowaboo,30
3,Zooloretto: The Dice Game,15
4,Zooloretto Junior,30


In [ ]:
#Problem 7: Are there games that don’t have a playtime specified?

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT  ?name WHERE {
  ?game a bgg:Game ;
        bgg:hasName ?name .
  FILTER NOT EXISTS {
    ?game bgg:hasMaxGameTime ?maxGameTime .
  }
}
LIMIT 50
""")
#
pretty(g.query(myquery))


""


In [ ]:
#Problem 8 List games with a geek rating above 7.5, and bind a quality label "high quality"
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT  ?name ?quality WHERE {
  ?game a bgg:Game ;
        bgg:hasName ?name;
        bgg:hasGeekRating ?rating.
        FILTER(?rating > 7.5)
        BIND("high quality" as ?quality)
}
LIMIT 50
""")
pretty(g.query(myquery))

myquery2 = g.update("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

INSERT  {
  bgg:Quality rdf:type rdfs:Class .
  bgg:High_Quality rdfs:subclassOf bgg:Quality.
  ?game bgg:hasQuality bgg:High_Quality .
} WHERE {
  ?game a bgg:Game ;
        bgg:hasGeekRating ?rating.
        FILTER(?rating > 7.5)
}
""")







In [ ]:
myquery3 = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT  ?name ?quality WHERE {
  ?game a bgg:Game ;
        bgg:hasName ?name.
  ?game bgg:hasQuality ?quality.
}
LIMIT 50
""")

print("After insert, we see the new resourcecs:\n\n")
pretty(g.query(myquery3))

After insert, we see the new resourcecs:




,name,quality
0,Trajan,https://raw.githubusercontent.com/susvej/bg_on...
1,Caverna: The Cave Farmers,https://raw.githubusercontent.com/susvej/bg_on...
2,A Game of Thrones: The Board Game (Second Edit...,https://raw.githubusercontent.com/susvej/bg_on...
3,Star Wars: X-Wing Miniatures Game,https://raw.githubusercontent.com/susvej/bg_on...
4,Descent: Journeys in the Dark (Second Edition),https://raw.githubusercontent.com/susvej/bg_on...
5,Lords of Waterdeep,https://raw.githubusercontent.com/susvej/bg_on...
6,War of the Ring (Second Edition),https://raw.githubusercontent.com/susvej/bg_on...
7,Terra Mystica,https://raw.githubusercontent.com/susvej/bg_on...
8,Robinson Crusoe: Adventures on the Cursed Island,https://raw.githubusercontent.com/susvej/bg_on...
9,Keyflower,https://raw.githubusercontent.com/susvej/bg_on...


In [ ]:
#Problem 10.pre what is the lowest and highest geek rating?
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT  (MAX(?rating) as ?maxrating) (MIN(?rating) as ?minrating)  WHERE {
  ?game a bgg:Game ;
        bgg:hasName ?name ;
        bgg:hasGeekRating ?rating.
}
""")
#
pretty(g.query(myquery))

,maxrating,minrating
0,8.52234,5.63619


In [ ]:
#Problem 10.pre what are the games with the lowest and highest geek rating?
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name ?rating WHERE {
  ?game a bgg:Game ;
        bgg:hasName ?name ;
        bgg:hasGeekRating ?rating.
}
ORDER BY DESC(?rating) LIMIT 5
""")
#
pretty(g.query(myquery))

,name,rating
0,Gloomhaven,8.52234
1,Pandemic Legacy: Season 1,8.49837
2,Through the Ages: A New Story of Civilization,8.32401
3,Twilight Struggle,8.21012
4,Terraforming Mars,8.17328


In [ ]:
#Find the name of the highest and lowest rated games

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name ?rating WHERE
{
  {
        # Subquery to get the maximum rating
        {
            SELECT (MAX(?r) AS ?targetRating) WHERE
            {
                ?g a bgg:Game ; bgg:hasGeekRating ?r .
            }
        }
        ?game a bgg:Game ;
              bgg:hasName ?name ;
              bgg:hasGeekRating ?rating .
        FILTER(?rating = ?targetRating)
  }
  UNION
  {
      # Subquery to get the minimum rating
        {
      SELECT (MIN(?r) as ?mintarget) WHERE
          {
          ?g rdf:type bgg:Game;
          bgg:hasGeekRating ?r.
          }
        }
      ?game rdf:type bgg:Game;
      bgg:hasName ?name;
      bgg:hasGeekRating ?rating.
      FILTER(?rating = ?mintarget)
    }
  }
""")
#
pretty(g.query(myquery))




,name,rating
0,Gloomhaven,8.52234
1,Crooks,5.63619


In [ ]:
#Find the name of the highest and lowest rated games

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?game
WHERE {
  {
    SELECT (MAX(?r) AS ?maxrating)
    WHERE {
      ?g bgg:hasRating ?r .
    }
  }
  ?game bgg:hasRating ?rating .
  FILTER(?rating = ?maxrating)
}
""")
#
pretty(g.query(myquery))


""


In [ ]:
#Problem 10: Bind a rating category based on rating thresholds: <0.6 "bad", 0.6-0.8 "ok", "0.8-1.0"
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?n ?rat ?opinion WHERE {
  ?game rdf:type bgg:Game;
    bgg:hasName ?n;
    bgg:hasGeekRating ?rat
  BIND(
    IF(?rat <=6, "bad",
      IF(?rat >6 && ?rat <=0.8, "ok", "good"))
    as ?opinion
  )
}
ORDER BY ?rat
LIMIT 10
""")
#
pretty(g.query(myquery))

,n,rat,opinion
0,Crooks,5.63619,bad
1,Stratego: The Lord of the Rings,5.6362,bad
2,Pecunia non olet,5.63622,bad
3,1829 Mainline,5.63625,bad
4,Onslaught,5.63625,bad
5,Kharkov,5.63629,bad
6,Hexemonia,5.63635,bad
7,Lancashire Railways,5.63647,bad
8,Richelieu,5.63655,bad
9,Moscow '41,5.63656,bad


In [ ]:
#Problem 10: Figure out the average game play time for all games and add a rating of "shorter" if it is below that, or "longer" if it is above

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name ?timerating WHERE {
  {
    SELECT (AVG(?time) as ?avgtime) WHERE {
      ?game rdf:type bgg:Game;
        bgg:hasMaxGameTime ?time
    }
  }
?game rdf:type bgg:Game;
bgg:hasName ?name;
bgg:hasMaxGameTime ?t2
BIND(
  IF(?t2 < ?avgtime, "short", "long")
  as ?timerating
)
}
""")
#
pretty(g.query(myquery))

,name,timerating
0,Die Macher,long
1,Elfenland,short
2,Hero of Weehawken,short
3,Seven Sisters,short
4,Cosmic Eidex,short
...,...,...
4994,Rolling Stock,long
4995,Swinging Jivecat Voodoo Lounge,short
4996,Pizza Theory,short
4997,Martian Dice,short


In [ ]:
#11: Count how many games each max player count supports (grouping) ("There are x max 2 player games, y max 3 player games etc")
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (COUNT(?game) as ?numgames) ?maxplayers WHERE {
  ?game rdf:type bgg:Game;
  bgg:hasMaxPlayers ?maxplayers
}
GROUP BY ?maxplayers
ORDER BY ASC(?maxplayers)
""")
#
pretty(g.query(myquery))

,numgames,maxplayers
0,25,0
1,77,1
2,955,2
3,64,3
4,1635,4
5,902,5
6,795,6
7,92,7
8,233,8
9,19,9


In [ ]:
#12: Give me 5 games whose name or category does not contain "War" (case-insensitive) but does have "soldier"
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name WHERE {
  ?g bgg:hasName ?name
  FILTER(!regex(?name, "war", "i") && regex(?name, "soldier", "i"))
}
LIMIT 5
""")
#
pretty(g.query(myquery))

,name
0,Soldier Kings


In [ ]:
#13: Find the average geek rating with 2 decimals per category and order it descending

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT
(ROUND(AVG(?rating) *100)/100 as ?average)
 (COUNT(?game) as ?numgames)
 ?cat WHERE {
  ?game rdf:type bgg:Game;
  bgg:hasGeekRating ?rating;
  bgg:hasCategory ?cat
}
GROUP BY ?cat
ORDER BY DESC(?average)
""")
#
pretty(g.query(myquery))

,average,numgames,cat
0,6.41,81,https://raw.githubusercontent.com/susvej/bg_on...
1,6.4,50,https://raw.githubusercontent.com/susvej/bg_on...
2,6.36,168,https://raw.githubusercontent.com/susvej/bg_on...
3,6.34,25,https://raw.githubusercontent.com/susvej/bg_on...
4,6.33,117,https://raw.githubusercontent.com/susvej/bg_on...
...,...,...,...
79,5.89,61,https://raw.githubusercontent.com/susvej/bg_on...
80,5.88,57,https://raw.githubusercontent.com/susvej/bg_on...
81,5.87,60,https://raw.githubusercontent.com/susvej/bg_on...
82,5.81,8,https://raw.githubusercontent.com/susvej/bg_on...


In [ ]:
#14: Find the average playtime for games with more than 4 max players
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (ROUND(AVG(?time)) as ?avgtime) WHERE {
  ?game rdf:type bgg:Game;
  bgg:hasMaxPlayers 4;
  bgg:hasMaxGameTime ?time
}
""")
#
pretty(g.query(myquery))

,avgtime
0,90


In [ ]:
#15: 8. Average rating per category, but only for groups with more than 20 games
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (AVG(?rating) as ?avgrating) (COUNT(?game) as ?numgames) ?category {
  ?game rdf:type bgg:Game;
     bgg:hasGeekRating ?rating;
     bgg:hasCategory ?category
}
GROUP BY ?category
HAVING (COUNT(?game) >20)
ORDER BY ASC(?numgames)
""")
#
pretty(g.query(myquery))


,avgrating,numgames,category
0,5.904596190476191,21,https://raw.githubusercontent.com/susvej/bg_on...
1,6.049693181818182,22,https://raw.githubusercontent.com/susvej/bg_on...
2,6.29452347826087,23,https://raw.githubusercontent.com/susvej/bg_on...
3,5.957355,24,https://raw.githubusercontent.com/susvej/bg_on...
4,5.9827832,25,https://raw.githubusercontent.com/susvej/bg_on...
...,...,...,...
71,6.236697480769241,520,https://raw.githubusercontent.com/susvej/bg_on...
72,6.308784187725635,554,https://raw.githubusercontent.com/susvej/bg_on...
73,6.169105598885797,718,https://raw.githubusercontent.com/susvej/bg_on...
74,5.971038670886091,948,https://raw.githubusercontent.com/susvej/bg_on...


In [ ]:
#16 10 Show mechanics used by at least 5 games, along with average play time.

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?mechanic (ROUND(AVG(?time)) as ?avgtime) (MAX(?time) as ?maxtime) WHERE {
  ?game rdf:type bgg:Game;
    bgg:hasMechanic ?mechanic;
    bgg:hasMaxGameTime ?time
}
GROUP BY ?mechanic
HAVING(COUNT(?game >5))
LIMIT 10
""")
#
pretty(g.query(myquery))


,mechanic,avgtime,maxtime
0,https://raw.githubusercontent.com/susvej/bg_on...,95,600
1,https://raw.githubusercontent.com/susvej/bg_on...,98,720
2,https://raw.githubusercontent.com/susvej/bg_on...,189,22500
3,https://raw.githubusercontent.com/susvej/bg_on...,61,1000
4,https://raw.githubusercontent.com/susvej/bg_on...,68,480
5,https://raw.githubusercontent.com/susvej/bg_on...,67,360
6,https://raw.githubusercontent.com/susvej/bg_on...,127,3000
7,https://raw.githubusercontent.com/susvej/bg_on...,115,720
8,https://raw.githubusercontent.com/susvej/bg_on...,205,12000
9,https://raw.githubusercontent.com/susvej/bg_on...,41,120


In [ ]:
#17 Find creators who have published more than 5 games and whose games have an average rating above 7.


myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?designer (AVG(?rating) as ?avgrating) (COUNT(?game) as ?numgames) WHERE {
  ?game rdf:type bgg:Game;
    bgg:hasCreator ?designer;
    bgg:hasGeekRating ?rating
}
GROUP BY ?designer
HAVING ((AVG(?rating) > 7) && (COUNT(?game) >5))
ORDER BY ?avgrating

""")
#
pretty(g.query(myquery))

,designer,avgrating,numgames
0,Vital Lacerda,7.1717,6
1,Corey Konieczka,7.227071428571429,7


In [ ]:
#18 Number of games per max playtime bucket (e.g., ≤30, 31–60, >60 min)
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (COUNT(?game) as ?numgames) ?bucket WHERE {
  ?game rdf:type bgg:Game;
     bgg:hasMaxGameTime ?maxtime
  BIND(IF(?maxtime <= 30, "less than 30",
        IF((?maxtime >30 && ?maxtime <=60), "less than 60", "longer than 60"))
  as ?bucket)
}
GROUP BY ?bucket
""")
#
pretty(g.query(myquery))

,numgames,bucket
0,1906,longer than 60
1,1585,less than 60
2,1508,less than 30


In [ ]:
#19: Find player count ranges (min+max) that occur more than once and their average rating.
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (CONCAT(STR(?minplayers), " to ",STR(?maxplayers)) as ?playerrange) (AVG(?rating) as ?avgrating) WHERE
{
  ?game rdf:type bgg:Game;
    bgg:hasMinPlayers ?minplayers;
    bgg:hasMaxPlayers ?maxplayers;
    bgg:hasGeekRating ?rating
}
GROUP BY ?minplayers ?maxplayers
""")
#
pretty(g.query(myquery))

,playerrange,avgrating
0,3 to 5,6.097763303167421
1,2 to 6,6.016111764705884
2,1 to 1,5.9058387012987
3,3 to 6,5.990852335025379
4,3 to 3,6.127341111111111
...,...,...
80,8 to 24,6.10707
81,5 to 34,5.70361
82,6 to 99,5.8217
83,5 to 12,5.64864


In [ ]:
#17 I want games that are likely to have to do with animals, so with any of these in the title
# animals ="animal, cat, dog, horse, cow, pig, sheep, goat, chicken, duck, rabbit, deer, bear, fox, wolf, lion, tiger, elephant, zebra, monkey, giraffe, kangaroo, camel, raccoon, squirrel, mouse, rat, hamster, guinea pig, donkey, turkey, peacock, frog, toad, snake, lizard, turtle, penguin, dolphin, whale, shark, octopus"
animals ="animal, cat, dog, horse, cow"

animalor = "|".join([a.strip() for a in animals.split(",")])
print(animalor)



animal|cat|dog|horse|cow


In [ ]:
#17 I want games that are likely to have to do with animals, so with any of these in the title animal, cat, dog, horse, cow

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name WHERE {
  ?game bgg:hasName ?name
  FILTER(regex(?name, "(^| )(animal|cat|dog|horse|cow)($s| )", "i"))
}
""")
#
pretty(g.query(myquery))

,name
0,Animal Upon Animal: Small and Yet Great!
1,Animal Upon Animal: Crest Climbers
2,The Really Nasty Horse Racing Game
3,Dog Royal
4,Cat Tower
5,Animal Upon Animal
6,Cat Box
7,Cat Lady
8,Battue: Storm of the Horse Lords
9,Felix: The Cat in the Sack


In [ ]:
#18 Find games without a category, if any
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name WHERE {
  ?game rdf:type bgg:Game
  FILTER NOT EXISTS {?game bgg:hasCategory ?category}
}
""")
#
pretty(g.query(myquery))

""


In [ ]:
#18 How many games have just 1,2,3 etc mechanics?
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT (COUNT(?game)as ?numgames) ?numech WHERE {
SELECT DISTINCT ?game (COUNT(?mech) as ?numech) WHERE {
 ?game rdf:type bgg:Game;
    bgg:hasMechanic ?mech
}
GROUP BY ?game
}
GROUP BY ?numech
ORDER BY ?numech
""")
#
pretty(g.query(myquery))

,numgames,numech
0,1042,1
1,1260,2
2,1239,3
3,756,4
4,375,5
5,179,6
6,92,7
7,29,8
8,16,9
9,6,10


In [ ]:
#19 Problem 1: Find 5 games that are either designed by "Reiner Knizia" or belong to the category "Strategy".

myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>
 SELECT ?gamename WHERE {
  {?game bgg:hasCreator "Reiner Knizia"; bgg:hasName ?gamename}
  UNION
  {?game bgg:hasCategory bgg:Abstract_Strategy; bgg:hasName ?gamename}
 }
LIMIT 5
""")
#
pretty(g.query(myquery))

,gamename
0,Carnac
1,Epigo
2,Tantrix
3,Casa Grande
4,Gyges


In [ ]:
#20  Games that are either in the category "Humor" or "Party_Game", but not both.
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name WHERE {
  {
   ?game bgg:hasCategory bgg:Humor; bgg:hasName ?name
   FILTER NOT EXISTS {?game bgg:hasCategory bgg:Party_Game}
   }
   UNION
  {
  ?game bgg:hasCategory bgg:Party_Game; bgg:hasName ?name
     FILTER NOT EXISTS {?game bgg:hasCategory bgg:Humor}

  }
}

""")
#
pretty(g.query(myquery))

,name
0,Dungeon Fighter
1,Farmageddon
2,Hex Hex
3,Nefarious
4,You're Bluffing!
...,...
346,Werewolf
347,Privacy
348,Tem-Purr-A
349,Cluzzle


In [ ]:
#22 List each bgg:Creator with their average rating and the number of games they designed.
# Only show those who designed more than 5 games.
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?creator (AVG(?rating) as ?avgrating) (COUNT(?game) as ?numgames) WHERE {
 ?game bgg:hasCreator ?creator;
   bgg:hasGeekRating ?rating
}
GROUP BY ?creator
HAVING (COUNT(?game) >5)
ORDER BY DESC(?numgames)

""")
#
pretty(g.query(myquery))

,creator,avgrating,numgames
0,Reiner Knizia,6.131383596491227,114
1,(Uncredited),6.018618615384616,65
2,Martin Wallace,6.333303061224489,49
3,Klaus Teuber,6.0992676315789485,38
4,Friedemann Friese,6.178309166666666,36
...,...,...,...
126,Chris Taylor (I),5.882173333333334,6
127,Chad Jensen,6.534006666666667,6
128,Vladimír Suchý,6.585326666666666,6
129,Joseph Miranda,5.77984,6


In [ ]:
#23 For each game, indicate whether its rating is above the average rating across all games.
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>
SELECT ?game ?rating ?ratinganalysis WHERE {
  {
    SELECT (AVG(?r) AS ?avgrating) WHERE {
      ?g bgg:hasGeekRating ?r .
    }
  }
  ?game bgg:hasGeekRating ?rating .
  BIND(IF(?rating >= ?avgrating, "over", "under") AS ?ratinganalysis)
}

""")
#
pretty(g.query(myquery))

,game,rating,ratinganalysis
0,https://raw.githubusercontent.com/susvej/bg_on...,7.25372,over
1,https://raw.githubusercontent.com/susvej/bg_on...,6.56289,over
2,https://raw.githubusercontent.com/susvej/bg_on...,5.74276,under
3,https://raw.githubusercontent.com/susvej/bg_on...,5.64146,under
4,https://raw.githubusercontent.com/susvej/bg_on...,5.64146,under
...,...,...,...
4994,https://raw.githubusercontent.com/susvej/bg_on...,5.85565,under
4995,https://raw.githubusercontent.com/susvej/bg_on...,5.77209,under
4996,https://raw.githubusercontent.com/susvej/bg_on...,5.68352,under
4997,https://raw.githubusercontent.com/susvej/bg_on...,6.07182,under


#RDF and OWL questions

###Challenge: Define GameDesigner as a class of individuals that have created at least one Game.

```
bgg:GameDesigner rdf:type owl:Class;
 owl:equivalentClass [
  rdf:type owl:Restriction;
  owl:onProperty bgg:hasCreated;
  owl:somValuesFrom :Game
 ]
```

### Model a class FamilyGame
... where a FamilyGame is any game that has as a category EITHER Party_Game or Trivia.

```
bgg:FamilyGame a owl:Class;
  owl:equivalentClass [
    owl:unionOf (
      [
        a owl:Restriction;
        owl:onProperty bgg:hasCategory;
        owl:hasValue bgg:Party_Game
      ]
      [
        a owl:Restriction;
        owl:onProperty bgg:hasCategory;
        owl:hasValue bgg:Trivia
      ]
    )

  ]
```

Or alternatively

```
bgg:FamilyGame a owl:Class;
  owl:equivalentClass [
   a owl:Restriction;
   owl:onProperty;
   owl:oneOf [
     owl:unionOf (bgg:Party_Game bgg:Trivia)
   ]
  ]
```

Compare when it is classes!
```
bgg:FamilyGame a owl:Class;
  owl:equivalentClass [
    a owl:Restriction;
    owl:onProperty bgg:hasCategory;
    someValuesFrom [
      owl:unionOf (bgg:Class1 bgg:Class2)
    ]

  ]
  ```

### Model a dice and card agme
That must have both the mechanic bgg:Dice_Rolling and bgg:Card_Drafting
```
bgg:DiceAndCardGame a owl:Class;
  owl:equivalentClass [
    owl:intersectionOf (
   [ a owl:Restriction;
   owl:onProperty bgg:hasMechanic;
   owl:hasValue bgg:Dice_Rolling
  ]
  [ a owl:Restriction;
    owl:onProperty bgg:hasMechanic;
   owl:hasValue bgg:Card_Drafting
  ]
    )
  ]
  ```

### Model KidFriendlyGame
which has to be from the Childrens_Game and be less than 30 minutes max

```
bgg:KidFriendlyGame a owl:Class;
 owl:equivalentClass [
  owl:intersectionOf (
  [
    a owl:Restriction;
    owl:onProperty bgg:Category;
    owl:hasValue bgg:Childrens_Game
  ]
  [
    a owl:Restriction;
    owl:onProperty bgg:MaxPlayTime
    owl:allValuesFrom [
      rdf:type owl:Datatype;
      owl:onDatatype xsd:integer;
      owl:withRestrictions
        [ (xsd:MaxInclusive 30)]

      )
    ]
  ]
  )
 ]

 ```

### Model HorrorCardGame
 which has to have categories bgg:Zombies or bgg:Horror and also have mechanic bgg:Card_Drafting

In [ ]:
```
bgg:HorrorCardGame a owl:Class;
 owl:equivalentClass [
     owl:intersectionOf (
     [a owl:Restriction;
      owl:onProperty bgg:hasCategory;
      owl:someValuesFrom
          [a owl:Class; owl:oneOf (bgg:Zombies bgg:Horror)]]
     [a owl:Restriction;
      owl:onProperty bgg:hasMechanic;
      owl:hasValue bgg:Card_Drafting]
     )
 ]

```

SyntaxError: invalid syntax (<ipython-input-39-da50aac62924>, line 1)

## Define a Class for “Games with exactly 3 distinct Mechanics

bgg:ThreeMechanics a owl:Class;
  owl:equivalentClass [
   a owl:Restriction;
   owl:onProperty bgg:hasMechanic;
   owl:onClass bgg:Mechanic
   owl:QualifiedCardinality "3"
  ]

##Model a Game Series with Versioning and Inheritance

```
bgg:BugHouseChessGame a owl:Class; subclassOf bgg:Game;
  owl:equivalentClass [
    olw:intersectionOf (
  [
   a owl:Restriction;
   owl:onProperty bgg:hasCreator;
   owl:hasValue [
    owl:Class
    owl:oneOf ("Paul", "Simon")]
  ]
  [a owl:Restriction;
   owl:onProperty bgg:hasCategory;
   owl:someValuesFrom [
    owl:oneOf(bgg:Abstract_Strategy)]
   [a owl:Restriction;
   owl:onProperty bgg:hasVersionNumber;
   owl:minCardinality 1;
   owl:maxCardinality 1]
   [a owl:Restriction;
   owl:onProperty bgg:hasPreviousVersion;
   owl:allValuesFrom bgg:Game]
    )]

bgg:14188 a bgg:BugHouseChessGame; bgg:hasCategory bgg:Abstract_Strategy ;
    bgg:hasCreator "Paul" ;
    bgg:hasGeekRating "5.90988" ;
    bgg:hasMaxGameTime 10 ;
    bgg:hasCategory bgg:Abstract_Strategy;
    bgg:hasVersionNumber 1.

```

##Model a class bgg:HistoricalGame where:

The game has a bgg:hasCategory property.

That category must be one of a set of specific individuals: bgg:Wargame, bgg:World_War_I, bgg:World_War_II

And you must enforce that these are the only categories allowed for this class.

Bonus: Add a constraint that it must also have at least one bgg:Mechanic tagged as bgg:Card_Drafting or bgg:Area_Control.
```
bgg:HistoricalGame a owl:Class;
owl:equivalentClass
[
  a owl:Class;
  owl:intersectionOf (
  [
   a owl:Restriction [
    owl:onProperty bgg:hasCategory;
    owl:someValuesFrom [
    a owl:Class;
     owl:oneOf (bgg:Wargame bgg:World_War_I bgg:World_War_II)
       ]
  ]
  [
        a owl:Restriction [
    owl:onProperty bgg:hasMechanic;
    owl:someValuesFrom [
      a owl:Class;
      owl:oneOf (bgg:Card_Drafting bgg:Area_Control)
      ]
    ]
  ]
  )
 ]
 ```

## Model two classes: bgg:WarGame bgg:PeaceGame

Such that:

A bgg:WarGame has a bgg:hasCategory that includes conflict-based themes.

A bgg:PeaceGame explicitly avoids conflict-based themes (e.g., no bgg:WWII, bgg:Battle, etc).
```
 bgg:WarGameType a owl:Class;
  owl:equivalentClass [
   a owl:Restriction;
   owl:onProperty bgg:hasCategory;
   owl:someValuesFrom [
   a owl:Class;
   owl:oneOf (bgg:World_War_I bgg:World_War_II)]
  ] .

  bgg:PeaceGameType a owl:Class;
   owl:equivalentClass [
   owl:complementOf bgg:WarGameType].

  ```

#Mutual Constraints on Properties
Create a class bgg:DuelGame where:

The game must have exactly two players.

The minimum and maximum players must both be 2.

And the game must not have the category bgg:Team_Game.

Trick: Use both owl:cardinality and owl:allValuesFrom to tightly constrain.

```
bgg:DuelGameType a owl:Class;
 owl:equivalentClass [
  a owl:Class;
  owl:intersectionOf (
    [
  a owl:Restriction;
  owl:onProperty bgg:hasMinPlayers;
  owl:hasValue 2
    ]
    [
  a owl:Restriction;
  owl:onProperty bgg:hasMaxPlayers;
  owl:hasValue 2
    ]
    [
  owl:ComplementOf
  a owl:Restriction;
  owl:onProperty bgg:hasCategory;
  owl:allValuesFrom [
  owl:complementOf bgg:TeamGame ]
    ]
  )
 ]

```

## Class Based on Absence of a Property
Model a class bgg:SoloGame such that:

A game belongs to it if it does not have any value for bgg:hasMaxPlayers greater than 1.

Also, the only allowed mechanic is bgg:Solo_Play.

Use owl:complementOf
```
bgg:SoloGameType a owl:Class;
  owl:equivalentClass [
    a owl:Class;
    owl:intersectionOf (
  [a owl:Restriction;
  owl:onProperty bgg:hasMechanic;
  owl:hasValue bgg:Solo_Play
  ]
  [a owl:Restriction;
  owl:onProperty bgg:hasMaxPlayers
  owl:hasValue 1

  ]
    )
  ]

```

## Conditional Modeling
Model a class bgg:KidsEducationalGame such that:

If the game has the category bgg:Children's_Game, then it must also have bgg:hasMechanic of type bgg:Memory.

Otherwise, there are no restrictions.

Trick: Requires modeling implication logic (hard to do directly in OWL, easier in SWRL or SHACL).


SWRL


```
bgg:hasMechanic(?g, bgg:Childrens_Game) ^
-> bgg:hasMechanic(?g, bgg:Memory)
```

Or we can put this in a shacl rule (I will not do that at this time)
```

CONSTRUCT {?game bgg:hasMechanic bgg:Memory} WHERE

{?game bgg:hasCategory bgg:Childrens_Game.
FILTER NOT EXISTS (?game bgg:hasMechanic bgg:Memory]
```

#Only These Values (Closed World Modeling)
Model a class bgg:ColorGame that:

Has a property bgg:hasColorTheme.

Must have only the following values: bgg:Red, bgg:Blue, or bgg:Yellow.

Must not have any other bgg:hasColorTheme.

Use owl:allValuesFrom and an owl:oneOf list carefully.

bgg:ColorGameType a owl:Class:
owl:equivalentClassOf [
 a owl:Restriction;
 owl:onProperty bgg:hasColorTheme;
 owl:allValuesFrom [
  a owl:Class;
  owl:oneOf(bgg:Red bgg:Blue bgg:Yellow)
 ]

]

## Create a class :NonStrategyGame that includes all games that are not strategy games.

Expected modeling:
Use owl:complementOf to define a class that is the logical complement of :StrategyGame.
```
bgg:StrategyGameType a owl:Class;
owl:subclassOf [
 a owl:Restriction;
 owl:onProperty bgg:hasCategory;
 owl:hasValue bgg:Strategy
]

bgg:NonStrategyGameTyoe o owl:Class;
owl:complementOf bgg:StrategyGame
```

## Functional property: each game has exactly one name
Problem:
Define :hasGameName as a property such that each game can have only one name.
                                                                                
                                                                              
```
bgg:hasGameName a owl:DatatypePropety, owl:FunctionalProperty;
 rdfs:domain bgg:Game;
 rdfs:range xsd:string;
```

OR, not quite the same, this says that

??? HOW RESTRICT wheels to max 4

??? WHEN TO USE equiv WHEN subclass?

```
bgg:Game a owl:Class;
 owl:equivalentClass [
  a owl:Restriction;
  owl:onProperty bgg:hasName;
  owl:maxCardinality 1
 ]
```

OR similar
```
bgg:GameShape a sh:Nodeshape;
  sh:Property [
  sh:path bgg:hasName;
  sh:maxCardinality 1
  ]

```

# limit the property hasWheels to exactly 4 individuals of type Wheel, regardless of the subject?

rdf/owl is class-first so we cannot do this without specifying some class as the subject. Options include: put this restriction on owl:Thing so that it is inherited by all subclasses.
```

```

or... do a shacl statement without a targetclass and with a qualifiedCardinality with sh:onClass Wheel

## Functional property: designers' year of birth

ransitive property: game expansions
Problem:
Create a property :expands that means one game is an expansion of another.
If A expands B, and B expands C, then A expands C. Insert this into the graph, with some data, and then write a sparql query that uses it

```
bgg:expands a owl:ObjectProperty, owl:TransitiveProperty;
 rdfs:domain bgg:Game;
 rdfs:range bgg:Game.
```


In [ ]:
#23 insert Games that expands each other inthe database.

sparql_update = """
PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX owl: <http://www.w3.org/2002/07/owl#>

INSERT {
  bgg:expands a owl:ObjectProperty, owl:TransitiveProperty ;
              rdfs:domain bgg:Game ;
              rdfs:range bgg:Game .

  bgg:Game1 a bgg:Game .
  bgg:Game2 a bgg:Game ; bgg:expands bgg:Game1 .
  bgg:Game3 a bgg:Game ; bgg:expands bgg:Game2 .
  bgg:Game4 a bgg:Game ; bgg:expands bgg:Game3 .
}
WHERE {}
"""
g.update(sparql_update)

## display the new data
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>
SELECT ?s ?expands ?o WHERE{
  ?s bgg:expands ?o.
  BIND("expands" as ?expands)
}

""")
#
pretty(g.query(myquery))


In [ ]:
##See all games that expands anothergame, including via intermediate steps
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>
SELECT DISTINCT ?game ?game2 WHERE {
  ?game rdf:type bgg:Game; bgg:expands+ ?game2
}
LIMIT 20

""")
#
pretty(g.query(myquery))

In [ ]:
##See all games that expands Game1 at some level
##Note inverse query sign of ^bgg:expands+
myquery = prepareQuery("""
    PREFIX bgg: <https://raw.githubusercontent.com/susvej/bg_ontology/>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX  owl:  <http://www.w3.org/2002/07/owl#>
    PREFIX  sh: <http://www.w3.org/ns/shacl#>
    PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>
SELECT DISTINCT ?game2 WHERE {
  bgg:Game1 ^bgg:expands+ ?game2
}
LIMIT 20

""")
#
pretty(g.query(myquery))

In [ ]:
LIM

## Complement of a union
Problem:
Model a class :NonFamilyOrPartyGame which includes all games that are not in the union of :FamilyGame or :PartyGame.

```
bgg:NonFamilyOrPartyGame a owl:Class ;
  owl:equivalentClass [
    a owl:Class ;
    owl:complementOf [
      a owl:Restriction ;
      owl:onProperty bgg:hasCategory ;
      owl:someValuesFrom [
        a owl:Class ;
        owl:oneOf (bgg:FamilyGame bgg:PartyGame)
      ]
    ]
  ] .
  ```

## Inverse functional property
Problem:
Define game id such that each code identifies exactly one game.

```
bgg:GameID a owl:Class;
owl:hasKey (bgg:idstring).

bgg:idstring a owl:DatatypeProperty;
rdfs:domain bgg:GameID;
rdfs:range xsd:string.

bgg:hasGameID a owl:ObjectProperty, owl:InverseFunctionalProperty;
rdfs:domain bgg:Game;
rdfs:range bgg:GameID.
```

## Use complement and datatype restriction
Problem:
Model a class :NotLongGames where games are defined as not having playtime > 120 minutes.

## Transitivity in part-of relationships
Problem:
Model a transitive :partOf property to represent how a mini-expansion is part of an expansion, which is part of a base game.

## Symmetric Property
Scenario:
Model a :coAccountHolder relationship between people.
```
bgg:coAccountHolder a owl:ObjectProperty, owl:SymmetricProperty;
  rdfs:domain foaf:Person;
  rdfs:range foaf:Person.
```

## Asymmetric Property
Scenario:
Model a :isParentCompanyOf property.

```
:isParentCompanyOf a owl:ObjectProperty, AsymmetricProperty;
rdfs:domain :Company;
rdfs:range :Company
```


## Irreflexive Property
Scenario:
Model a :defeated property in a game setting.

Challenge:
A player cannot defeat themselves.

```
:defeated a owl:ObjectProperty; IrreflexiveProperty;
  rdfs:domain foaf:Person;
  rdfs:range foaf:Person.

```